Enable variable modifications to layer changes on the same variable

douglasjacobsen opened the pull request:
This merge allows modifiers that modify the same variable to layer their changes onto the modified variable.

Previously, only the last modifier to modify a variable would have it's modification stick.

douglasjacobsen committed:
Enable variable modifications to layer changes on the same variable

This commit allows modifiers that modify the same variable to layer
their changes onto the modified variable.

Previously, only the last modifier to modify a variable would have it's
modification stick.

linsword13 commented on lib/ramble/ramble/modifier.py:
```diff
@@ -255,10 +255,13 @@ def modded_variables(self, app):
                 # var_str = app.expander.expansion_str(var)
                 # prev_val = app.expander.expand_var(var_str)
                 prev_val = app.variables[var]
+                if var in extra_vars:
```
nit: maybe using the single-assignment-form is clearer?

```
if var in extra_vars:
    prev_val = extra_vars[var]
else:
    prev_val = app.variables[var]
```

douglasjacobsen replied:
Yeah, good point. Fixed.

linsword13 reviewed (commented):
LGTM. Do you think it'd be good to add a test for it?

douglasjacobsen committed:
Add a test for layered variable modifications

douglasjacobsen committed:
Add space separator to APS modifier

douglasjacobsen commented:
I added a test, and updated existing modifiers to include the space separator as part of their variable modifications (given I removed the auto-injected space)

linsword13 commented:
> I added a test, and updated existing modifiers to include the space separator as part of their variable modifications (given I removed the auto-injected space)

Do you feel having the separator auto-injected would be less error-prone than relying on modifier author to remember adding the space?

Perhaps a more general approach is to have the directive to accept a `template_func` (`(prev_value, new_mod) => new_value`)? (More just for a discussion, no need to add to the current PR though.)

douglasjacobsen commented:
Yes, I actually have an internal task tracking this which I think is a high priority.

Injecting a space manually is definitely a hack for the time being, and we should have a nicer interface for this.
